File: Python/Quicksort.py
"""
Quicksort
We divide our unsorted list in 2 smaller unsorted lists using a pivot-element.
On each of these smaller lists we call again the quicksort algorithm.
Repeat until the lists are of the length 1 or 0.
Worst case runtime O(n^2)
"""


def quick_sort(arr):

    def _partition(array, begin, end):
        pivot = begin
        for i in range(begin + 1, end + 1):
            if array[i] <= array[begin]:  # We increment i until we locate a value that is smaller than the pivot value
                pivot += 1
                array[i], array[pivot] = array[pivot], array[i]  # then we swap the smaller value with our pivot value
        array[pivot], array[begin] = array[begin], array[pivot]
        return pivot  # return the index of our pivot-element

    def _sort(_arr, start, end):
        if start >= end:
            return
        pivot = _partition(_arr, start, end)
        _sort(_arr, start, pivot - 1)
        _sort(_arr, pivot + 1, end)
    _sort(arr, 0, len(arr) - 1)  # our pivot is the first element
    return arr


if __name__ == "__main__":
    test_arr = [
        3,
        1,
        7,
        4,
        12,
        9,
        11,
        102
    ]
    print(quick_sort(test_arr))

File: C++/Quicksort.cpp
#include <iostream>
#include <vector>

#include "Randomdata.cpp"

void swap(int *a, int *b){
    // swap two values in an array/ vector
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(std::vector<int> &arr,int start,int end){
    int pivot = start;
    for (int i = start + 1; i <= end; i++){
        if (arr[i] <= arr[start]){
            pivot += 1;
            swap(&arr[i], &arr[pivot]);
        }
    }
    swap(&arr[pivot], &arr[start]);
    return pivot;
}

void quickSort(std::vector<int> &arr, int start=0, int end=-2){
    if (end == -2){
        quickSort(arr, 0, arr.size() - 1);
    }
    else if (start < end){
        int pivot = partition(arr, start, end);
        quickSort(arr, start, pivot - 1);
        quickSort(arr, pivot + 1, end);
    }
}

void printArray(const std::vector<int>& arr){
    for (auto const& v: arr){
        std::cout << v << " | ";
    }
}

int main(){
    std::vector<int> arr = {12, 321, 12, 1, 5, 34, 27};
    quickSort(arr);
    printArray(arr);

    std::cout << "\n\n";
    std::vector<int> bench = randomData(5000);
    printArray(bench);
    std::cout << "\n\n";
    quickSort(bench);
    printArray(bench);

    return 0;
}